tour.md

```diff
@@ -1,730 +1,493 @@
+if true break
+if true continue
+if true for ...
+if true while ...
+if true loop ...
+```
+
+Loops (for, while, loop), continue/break
+
+```rust
+for item in iterator { /*...*/ }
+
+while condition() { /*...*/ }
+loop { /*...*/ }
+
+for i in 0..10 {
+  print(i)
+}
+
+v := 10
+while v >= 0 {
+  print(v)
+  v -= 1
+}
+
+v := 0
+loop {
+  if v == 5 break
+  v += 1
+}
+```
+
+Functions (generics, variadics, return)
+
+```rust
+fn name(a: A, b: B, c: C) {/*...*/}
+fn name[T](a: A, b: B, c: C) -> T {/*...*/}
+fn name[T, E](a: A, b: B, c: C) -> T where T: Bound {/*...*/}
+
+fn name[T, E](a: A, b: B, c: C) -> T
+  where
+    A: IA,
+    B: IB,
+    C: IC
+{
+  /*...*/
+}
+
+// anonymous function
+square := \x {x*x}
+
+fn fib(n: int): int {
+  if n < 2 { n }
+  else n * fib(n - 1)
+}
+
+// return may be used to return early
+fn test() {
+  if something() { return }
+  print("yo")
+}
+
+fn split(s: string, sep: string) -> [string] {
+  out := [];
+  start := 0;
+  end := 0;
+
+  for ch in s {
+    if ch == sep {
+      out.push(s[start..end])
+      start = end;
+    } else {
+      end += 1;
+    }
+  }
+
+  if current.len > 0 {
+    out.push(current);
+  }
+
+  out
+}
+
+print(split("a,b,c", ","))
+```
+
+Classes, traits
+
+```rust
+trait Default {
+  fn default() -> Self;
+}
+
+trait Iterate {
+  type Iter: Iterator;
+  fn iter(self) -> Iter;
+}
+
+trait Iterator {
+  type Item;
+  fn next(self) -> Item?;
+}
+
+class Node<T> {
+  value: T;
+  next: Node[T]?;
+}
+
+class List[T] {
+  @head: Node[T]?;
+
+  fn new() -> Self {
+    List(head: null);
+  }
+
+  fn prepend(self, value: T) {
+    next := self.head;
+    self.head = Node(value);
+    self.head.next = next;
+  }
+
+  impl Iterate {
+    type Iter = ListIter[T];
+    fn iter(self) -> Iter {
+      ListIter(node: self.head)
+    }
+  }
+}
+
+class ListIter[T] {
+  node: Node[T];
+
+  impl Iterator {
+    type Item = Node[T];
+    fn next(self) -> Item? {
+      node := self.node;
+      self.node = node.next;
+      node.value
+    }
+  }
+}
+
+list := List.new();
+
+for i in 0..10 {
+  list.prepend(i);
+}
+
+for val in list {
+  print(val);
+}
+```
+
+Exceptions
+
+```rust
+class A {}
+class B {}
+class C {}
+
+fn f() {
+  throw A()
+}
+
+v := try f() {
+  A => ...,
+  B => ...,
+  C => ...,
+}
+
+try {
+  a();
+  b();
+  c();
+} catch {
+  A => ...
+  B => ...
+  C => ...
+}
+```
+
+Modules
+
+```rust
+import module
+import module.A
+import module.{A, B}
+
+import nested.module.A
+import nested.module.{A, B}
+
+import {
+  a.{A, B},
+  b.{C, D},
+  nested.{
+    d.{E, F},
+    f.{G, H},
+  }
+}
+
+// you can export functions, classes, enums, and types
+export fn a() {}
+export class B {}
+export enum C {}
+export type D {}
+
+// exports may also be grouped into a block export
+fn e() {}
+traits F
+class G {}
+// symbols exported this way may be renamed using `as`
+export { e as g, F as E, G as F }
+
+// module imports may be arbitrarily re-ordered by the runtime, which means that

```

Comment:
Are you sure about `arbitrarily`? It's more like the independent subgraphs of the DAG can be executed concurrently. In the examples below, `foo` must and will be executed before `first` and `second`, but `first` and `second` may be executed concurrently since they don't depend on one-another.

Reply:
If we eventually turn this into some official tour posted online somewhere, I insist on the wording `arbitrary re-ordering`, because that gives you the least guarantees. The modules in the `concurrency` section are straightforward, but imagine adding just one level of depth to the DAG, where for each module you import two other modules, and then also import one or more of those bottom-most modules in `foo`:
```
    root -----.
   /    \     |
  a      b    |
 / \    / \   |
c   d  e   f<-|
```
In this case, the import order may be
```
root (f, a (c, d), b (e, f))
```
or it may be
```
root (a (c, d), b (e, f), f)
```

Reply:
Okay, I agree, it's beneficial from the implementation perspective. We shouldn't encourage users to introduce important side effects to the top-level code anyway.